explorer/indexer/app/db_test.go

```diff
@@ -0,0 +1,190 @@
+package app_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/omni-network/omni/explorer/db"
+	"github.com/omni-network/omni/explorer/db/ent"
+	"github.com/omni-network/omni/explorer/db/ent/block"
+	"github.com/omni-network/omni/explorer/db/ent/msg"
+	"github.com/omni-network/omni/explorer/db/ent/receipt"
+	"github.com/omni-network/omni/explorer/indexer/app"
+	"github.com/omni-network/omni/lib/xchain"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+)
+
+type results struct {
+	blocks   []ent.Block
+	messages []*ent.Msg
+	receipts []*ent.Receipt
+}
+
+type prerequisite func(t *testing.T, ctx context.Context, client *ent.Client) results
+
+func TestDbTransaction(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	type want struct {
+		BlockCount        int
+		MsgCount          int
+		ReceiptCount      int
+		StrayReceiptCount int
+		StrayMessageCount int
+	}
+
+	tests := []struct {
+		name          string
+		prerequisites prerequisite // These functions create entries on our db before the evaluation
+		want          want
+	}{
+		{
+			name: "insert_block_with_msgs_and_receipts",
+			prerequisites: func(t *testing.T, ctx context.Context, client *ent.Client) results {
+				t.Helper()
+				tx, err := client.BeginTx(ctx, nil)
+				require.NoError(t, err)
+
+				sourceChainID := uint64(1)
+				destChainID := uint64(2)
+
+				//client.XProviderCursor.Create().SetChainID(sourceChainID).SetHeight(0).SaveX(ctx)
+				//client.XProviderCursor.Create().SetChainID(destChainID).SetHeight(0).SaveX(ctx)
+
+				blockHash := common.Hash([32]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 30})
```

Comment:
you can fuzz all of these values, to avoid writing mock hard-coded values. For example:
```go
import (
     fuzz "github.com/google/gofuzz"
)

func example() {
     var blockHash common.Hash
     fuzz.New().NilChance(0).Fuzz(&blockHash)
}
```

Reply:
also `tutil.RandomHash()`